docs/design/baremetal/networking-infrastructure.md

```diff
@@ -70,7 +70,101 @@ DNS requirements internal to the cluster as possible, leaving only a
+discoverable by other hosts in the cluster.  It will be run by the
+`machine-config-operator`, once the following PR is complete:
+
+https://github.com/openshift/machine-config-operator/pull/795
+
+### coredns-mdns
+
+https://github.com/openshift/coredns-mdns/
+
+The `mdns` plugin for `coredns` was developed to perform DNS lookups
+based on discoverable information from mDNS.  This plugin will resolve both the
+`etcd-NNN` records, as well as the `_etcd-server-ssl._tcp.` SRV record.
+
+The list of `etcd` hosts included in the SRV record is based on the list of
+control plane nodes currently running.
+
+The IP addresses that the `etcd-NNN` host records resolve to comes from the
+mDNS advertisement sent out by the `mdns-publisher` on that control plane noded
+
+### etcd clustering with DNS SRV
+
+This section is not `baremetal` platform specific, but it’s important to
+understand as the DNS automation put in place is to support this process.
+
+`etcd` includes support for using DNS as a discovery mechanism.  The `etcd`
+documentation for this is here: https://github.com/openshift/etcd/blob/master/Documentation/op-guide/clustering.md#dns-discovery
+
+OpenShift makes use of this DNS based discovery mechanism to form the `etcd`
+cluster.  You can see the following if you look at one of the `etcd` member
+pods in a running cluster:
+
+```
+$ oc get pod etcd-member-master-0 -n openshift-etcd -o yaml | grep -A 1 "exec etcd"
+      exec etcd \
+        --discovery-srv ostest.test.metalkube.org \
+```
+
+This static pod definition is generated by the `machine-config-operator` from
+this [template](https://github.com/openshift/machine-config-operator/blob/master/templates/master/00-master/_base/files/etc-kubernetes-manifests-etcd-member.yaml).
+
+### DNS Resolution During Bootstrapping
+
+One of the of virtual IP addresses required by the `baremetal` platform is used
+for our self-hosted internal DNS - the “DNS VIP”.  The location of the DNS VIP
+is managed by `keepalived`, similar to the management of the API VIP.
+
+The control plane nodes are configured to use the DNS VIP as their primary DNS
+server.  The DNS VIP resides on the bootstrap host until the control plane
+nodes come up, and then it will move to one of the control plane nodes.
+
+`coredns-mdns` on the bootstrap node is configured a little bit differently
+than `coredns-mdns` on the control plane nodes.  On the bootstrap node, it is
+configured with a minimum number of hosts for the `etcd` SRV record based on
+the number of control plane nodes requested in `install-config.yaml`.
+
+This configuration is important.  It ensures that whenever `etcd` resolves this
+DNS SRV record to form the cluster, all of the expected control plane hosts are
+present in the results.
+
+### DNS Resolution Post-Install
+
+After the bootstrap process completes, the DNS VIP moves to one of the control
+plane nodes.  All control plane nodes will continue to use this DNS VIP as
+their primary DNS server.
```

Comment:
Should we mention how that's done, e.g via a NM prepender script, or is that documented in the MCO integration?

Reply:
++ need to cover dns-vip-prepender

Reply:
Good point.  There's no documentation of the MCO integration, though.  I envision this document being the high level docs that discuss how all the pieces together.  Detailed documentation of the specific MCO pieces should perhaps be in the MCO repository.  That's where I'd expect to document the exact piece that performs this task.

Once the MCO docs exist, we should be linking to them from this document.

What do you think?